lib/Language/Souffle/Class.hs

```diff
@@ -0,0 +1,251 @@
+--
+-- @
+-- data Path = Path  -- Handle for the datalog program
+--
+-- instance Program Path where
+--   type ProgramFacts Path = '[Edge, Reachable]
+--   factName = const "path"
+-- @
+class Program a where
+  -- | A type level list of facts that belong to this program.
+  --   This list is used to check that only known facts are added to a program.
+  type ProgramFacts a :: [Type]
+
+  -- | Function for obtaining the name of a Datalog program.
+  --   This has to be the same as the name of the .dl file (minus the extension).
+  --
+  -- It uses a 'Proxy' to select the correct instance.
+  programName :: Proxy a -> String
+
+-- | A typeclass for data types representing a fact in datalog.
+class Marshal.Marshal a => Fact a where
+  -- | Function for obtaining the name of a fact
+  --   (has to be the same as described in the Datalog program).
+  --
+  -- It uses a 'Proxy' to select the correct instance.
+  --
+  -- Example usage:
+  --
+  -- @
+  -- instance Fact Edge where
+  --   factName = const "edge"
+  -- @
+  factName :: Proxy a -> String
+
+---- | Helper typeclass for collecting facts into a container-like structure.
+----   The order of returned facts is unspecified for performance reasons.
+----   Only used internally.
+--class CollectFacts m c where
+--  type CollectFactsIterator m c :: Type
+--  collectFacts :: Marshal.Marshal a
+--               => Int
+--               -> CollectFactsIterator m c
+--               -> IO (c a)
+
+-- | A mtl-style typeclass for Souffle-related actions.
+class Monad m => MonadSouffle m where
+  type Handler m :: Type -> Type
+  {- | Initializes a Souffle program.
+
+     The action will return 'Nothing' if it failed to load the Souffle program.
+     Otherwise it will return a 'Handle' that can be used in other functions
+     in this module.
+  -}
+  init :: Program prog => prog -> m (Maybe (Handler m prog))
+
+  -- | Runs the Souffle program.
+  run :: Handler m prog -> m ()
+
+  -- | Sets the number of CPU cores this Souffle program should use.
+  setNumThreads :: Handler m prog -> Word64 -> m ()
+
+  -- | Gets the number of CPU cores this Souffle program should use.
+  getNumThreads :: Handler m prog -> m Word64
+
+  -- | Load all facts from files in a certain directory.
+  loadFiles :: Handler m prog -> FilePath -> m ()
+
+  -- | Write out all facts of the program to CSV files
+  --   (as defined in the Souffle program).
+  writeFiles :: Handler m prog -> m ()
+
+  -- | Returns all facts of a program. This function makes use of type inference
+  --   to select the type of fact to return.
+  getFacts :: (Fact a, ContainsFact prog a)
+           => Handler m prog -> m [a]
+
+  -- | Searches for a fact in a program.
+  --   Returns 'Nothing' if no matching fact was found; otherwise 'Just' the fact.
+  --
+  --   Conceptually equivalent to @List.find (== fact) \<$\> getFacts prog@, but this operation
+  --   can be implemented much faster.
+  findFact :: (Fact a, ContainsFact prog a, Eq a)
```

Comment:
`findFact` wasn't clear, and for the interpeter it needed an extra `Eq` constraint.

Reply:
What isn't clear about it? Too little documentation? It's one of the later things I added, since the compiled version has a much quicker implementation than the conceptual implementation.

And that's funny, I didn't even realize I didn't need an `Eq` constraint for the compiled version until now (all happens in C++). 
I think it makes sense to add it, given that the docs mention `==` from `Eq` already.

Reply:
> What isn't clear about it? Too little documentation? 

I had the surprise of 'oh wait why there isn't an Eq constraint here?' for a moment, and after I realized probably the souffle API does the equality checking. The only problem I can see with adding the Haskell Eq check if the Eq check on the haskell side differs from the internal C++ souffle eq check. On the other hand nothing stops us creating a different `a` wrapped in a `Just` than the one we passed in. :)

Reply:
From what I've seen, it checks the equality in the same way as Haskell; each field of a fact (internally called a tuple in Souffle) should be same equal to the subfields of the fact we are looking for. And yes, this is why I didn't need the `Eq` constraint there.

How would you create a different `a` then? We know nearly nothing about it, except that it implements a few typeclass (but none allow creating a fact out of thin air). One of the reasons why I love parametric polymorphism so much 😉. 

I preferred returning `Maybe a` over `Bool` here since it feels much more high level (no boolean blindness!) and it allows using functions related to `Maybe` instead of tons of if statements 😄

Reply:
Different Eq implementation. I meant the other way around. If I create a weird Eq instance for my fact type on the Haskell side, switching from Interpreted to Compiled can cause some surprises, but I'll get my well deserved punishment! :D

Boolean blindness totally agree to that! :)